### PR TITLE
[Merged by Bors] - chore(SetTheory/Game/Nim): redefine `grundyValue` as an infimum

### PR DESCRIPTION
**Pull request by vihdzp**
We redefine `grundyValue` as an infimum instead of using `Ordinal.mex`, and we give it a better API.

This is a preliminary step before deprecating `Ordinal.mex` altogether - there's no point in using it since we can just use `Inf sᶜ` instead.

---

- [x] depends on: #16875

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/SetTheory/Game/Nim.lean [hunk 1]
```diff
@@ -206,30 +206,49 @@ theorem nim_add_fuzzy_zero_iff {o₁ o₂ : Ordinal} : nim o₁ + nim o₂ ‖ 0
...
 theorem grundyValue_ne_left {G : PGame} (i : G.LeftMoves) :
     grundyValue (G.moveLeft i) ≠ grundyValue G := by
-  conv_rhs => rw [grundyValue_eq_mex_left]
-  apply ne_mex _ i
+  conv_rhs => rw [grundyValue_eq_sInf_left]
+  have := csInf_mem (compl_nonempty_of_small (Set.range fun i => grundyValue (G.moveLeft i)))
+  rw [Set.mem_compl_iff, Set.mem_range, not_exists] at this
+  exact this _
+
+theorem le_grundyValue_of_subset_left {G : PGame} {o : Ordinal}
```
Comment (YaelDillies) [inline Mathlib/SetTheory/Game/Nim.lean:233]:
```suggestion
theorem le_grundyValue_of_Iio_subset_moveLeft {G : PGame} {o : Ordinal}
```
or something like that?

File: Mathlib/SetTheory/Game/Nim.lean [hunk 2]
```diff
@@ -206,30 +206,49 @@ theorem nim_add_fuzzy_zero_iff {o₁ o₂ : Ordinal} : nim o₁ + nim o₂ ‖ 0
...
 
 theorem exists_grundyValue_left_of_lt {G : PGame} {o : Ordinal} (h : o < grundyValue G) :
     ∃ i, grundyValue (G.moveLeft i) = o := by
-  rw [grundyValue_eq_mex_left] at h
-  exact exists_of_lt_mex h
+  rw [grundyValue_eq_sInf_left] at h
+  by_contra ha
+  exact h.not_le (csInf_le' ha)
+
+theorem grundyValue_le_of_forall_left {G : PGame} {o : Ordinal}
```
Comment (YaelDillies) [inline Mathlib/SetTheory/Game/Nim.lean:245]:
```suggestion
theorem grundyValue_le_of_forall_moveLeft {G : PGame} {o : Ordinal}
```
similarly

File: Mathlib/SetTheory/Game/Nim.lean [hunk 3]
```diff
@@ -274,66 +293,75 @@ theorem grundyValue_star : grundyValue star = 1 :=
 theorem grundyValue_neg (G : PGame) [G.Impartial] : grundyValue (-G) = grundyValue G := by
   rw [grundyValue_eq_iff_equiv_nim, neg_equiv_iff, neg_nim, ← grundyValue_eq_iff_equiv_nim]
 
-theorem grundyValue_eq_mex_right (G : PGame) [G.Impartial] :
-    grundyValue G = Ordinal.mex.{u, u} fun i => grundyValue (G.moveRight i) := by
+theorem grundyValue_eq_sInf_right (G : PGame) [G.Impartial] :
```
Comment (YaelDillies) [inline Mathlib/SetTheory/Game/Nim.lean:296]:
```suggestion
theorem grundyValue_eq_sInf_moveRight (G : PGame) [G.Impartial] :
```

File: Mathlib/SetTheory/Game/Nim.lean [hunk 4]
```diff
@@ -274,66 +293,75 @@ theorem grundyValue_star : grundyValue star = 1 :=
...
+    grundyValue G = Ordinal.mex.{u, u} fun i => grundyValue (G.moveRight i) :=
+  grundyValue_eq_sInf_right G
+
 theorem grundyValue_ne_right {G : PGame} [G.Impartial] (i : G.RightMoves) :
     grundyValue (G.moveRight i) ≠ grundyValue G := by
-  conv_rhs => rw [grundyValue_eq_mex_right]
-  apply ne_mex _ i
+  convert grundyValue_ne_left (toLeftMovesNeg i) using 1 <;> simp
+
+theorem le_grundyValue_of_subset_right {G : PGame} [G.Impartial] {o : Ordinal}
```
Comment (YaelDillies) [inline Mathlib/SetTheory/Game/Nim.lean:313]: Same

File: Mathlib/SetTheory/Game/Nim.lean [hunk 5]
```diff
@@ -274,66 +293,75 @@ theorem grundyValue_star : grundyValue star = 1 :=
...
-  exact exists_of_lt_mex h
-
--- Todo: this actually generalizes to all ordinals, by defining `Ordinal.lxor` as the pairwise
--- `Nat.xor` of base `ω` Cantor normal forms.
+  rw [← grundyValue_neg] at h
+  obtain ⟨i, hi⟩ := exists_grundyValue_left_of_lt h
+  use toLeftMovesNeg.symm i
+  rwa [← grundyValue_neg, ← moveLeft_neg']
+
+theorem grundyValue_le_of_forall_right {G : PGame} [G.Impartial] {o : Ordinal}
```
Comment (YaelDillies) [inline Mathlib/SetTheory/Game/Nim.lean:326]: Same

**Review by YaelDillies — APPROVED** [review]
Thanks!

maintainer merge

**Comment by semorrison** [comment]
bors merge

**Comment by bryangingechen** [comment]
This fell off the queue when bors crashed:
bors r+

In case something bad happens again:
bors d+

**Pull request closed without merging**